Feature/export

zwan0202 opened the pull request:
Why is this change needed?
goal of this PR is to add functionality so we can export project to EDL
What did you change?
added export functionality and modified the mock type for Transcription

itsMoxMox committed:
Added initial files

Mhudlark committed:
Add test commit

itsMoxMox committed:
Implemented basic functionality for EDL Export

zwan0202 committed:
Update export '.edl' file into 'export_files' folder

zwan0202 committed:
Add helper functions into 'utils.ts'

zwan0202 committed:
Refactor 'exportEDL' to take 'ProjectBase' as input

zwan0202 committed:
Merge branch 'develop' into feature/export

zwan0202 committed:
Update 'Transcription' type in 'sharedTypes'

zwan0202 committed:
Remove export folder

zwan0202 committed:
Fix .gitignore and .eslintignore file

itsMoxMox reviewed (changes requested):
Will just need to address the merge conflict where dev's sharedTypes references a now outdated mockup for Transcription.

itsMoxMox committed:
Merge branch 'develop' into feature/export

zwan0202 committed:
Resolve merge conflict sharedTypes references

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,49 @@
+import { Project, Transcription } from '../sharedTypes';
+import { padZeros, integerDivide, writeFile, mkdir } from './util';
+
+const secondToTimestamp = (num: number) => {
```
For consistency let's make sure all function signatures have a return type, like this:

`const secondToTimestamp: (num: number) => string = (num) => { ... }`

zwan0202 replied:
Right, I'll keep that in mind

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,49 @@
+import { padZeros, integerDivide, writeFile, mkdir } from './util';
+
+const secondToTimestamp = (num: number) => {
+  return `${padZeros(integerDivide(num, 3600), 2)}:
```
Might want to abstract this a little further, up to you what you prefer

e.g. `[3600, 60, 1, 0.01].map(mult => padZeros(integerDivide(num, mult), 2)).join(':')`

That said it's kind of hard to understand as is, maybe add a comment describing what the timestamp looks like, and/or constants for the numeric values (to avoid 'magic numbers')

zwan0202 replied:
Thanks, that looks much cleaner

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,49 @@
+  try {
+    const { words } = transcription;
+    const entries = words.length;
+    for (let i = 0; i < entries; i++) {
```
Could be cleaner to do this with `map` since every iteration of the loop is just producing a new line of output

zwan0202 replied:
Got it

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,49 @@
+      output += `${edlEntry}\t${editStart}\t${editEnd}\n* FROM CLIP NAME: sample\n\n`;
+    }
+  } catch (e) {
+    console.error(e);
```
What errors are you catching here?

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,49 @@
+  }
+};
+
+const exportFuncs = {
```
What's wrong with just exporting default exportEDL?

chloebrett commented on src/main/util.ts:
```diff
@@ -16,3 +17,29 @@ if (process.env.NODE_ENV === 'development') {
+  try {
+    writeFileSync(filePath, data, { flag: 'w' });
+  } catch (err) {
+    console.error(err);
```
Hm, where is this being used and would it be better to have that function throw? We shouldn't normally get errors writing a file and if we do we definitely don't want to continue like nothing happened

hhen0005 replied:
I reckon Patrick is right. Fail fast. 
I have added a ticket to our team backlog to display an error to the user when export fails.

chloebrett commented on src/sharedTypes.ts:
```diff
@@ -1,10 +1,17 @@
+
+export type Word = {
+  word: string;
+  start_time: number;
```
Camelcase attributes

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,49 @@
+};
+
+const constructEDL = (title: string, transcription: Transcription) => {
+  let output = `TITLE: ${title}\nFCM: NON-DROP FRAME\n\n`;
```
Can you add a link to your write up (I think you made a write up?) on what the structure of an EDL is like? A G Drive link ideally

zwan0202 replied:
Not me, I think Max wrote that

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,49 @@
+          ${padZeros(integerDivide(num, 0.01), 2)}`;
+};
+
+const constructEDL = (title: string, transcription: Transcription) => {
```
This is a good candidate for some unit testing, fine to do this in a later PR, would you be confident in adding some tests with jest?

zwan0202 replied:
I can try

chloebrett replied:
Great, feel free to reach out (to myself or anyone else familiar with jest) if you need a hand

chloebrett commented on src/sharedTypes.ts:
```diff
@@ -1,10 +1,17 @@
 export interface ProjectBase {
   schemaVersion: number;
   name: string;
+  savePath: string;
```
Not sure if this should be stored on the project entity itself, or just part of the state that initialises when we open a file (since the project file could be moved), but this is fine for now

hhen0005 replied:
Might be useful to store the prev location of export. At the very least, maybe rename to `exportPath` or `exportDefaultPath` rather than `savePath`. Might be confusing is all.

chloebrett reviewed (changes requested):
Looking good! Your yarn.lock is a little messed up, not sure why - maybe discard your changes to it and re-run yarn

Also, we don't want to have a package-lock.json, that's a symptom of running `npm install` which we don't need to do since we are using yarn.

I would suggest
1. Delete package-lock.json
2. Reset your changes to yarn.lock: `git checkout develop; git pull; git checkout feature/export; git checkout develop -- yarn.lock`
3. Re-run yarn
4. If that doesn't work, delete your node_modules directory and try steps 2-3 again

Everything else is pretty nice, just added a few comments for places you could simplify things and some minor style stuff

hhen0005 reviewed (commented):
Waiting on suggested changes.

zwan0202 committed:
Fix suggested changes

zwan0202 commented:
> Looking good! Your yarn.lock is a little messed up, not sure why - maybe discard your changes to it and re-run yarn
> 
> Also, we don't want to have a package-lock.json, that's a symptom of running `npm install` which we don't need to do since we are using yarn.
> 
> I would suggest
> 
> 1. Delete package-lock.json
> 2. Reset your changes to yarn.lock: `git checkout develop; git pull; git checkout feature/export; git checkout develop -- yarn.lock`
> 3. Re-run yarn
> 4. If that doesn't work, delete your node_modules directory and try steps 2-3 again
> 
> Everything else is pretty nice, just added a few comments for places you could simplify things and some minor style stuff

I deleted node_modules and redo the steps 2-3, not sure why all the `resolved` path in my `yarn.lock` is `npmjs` instead of `yarnpkg`

chloebrett commented on src/main/export.ts:
```diff
@@ -0,0 +1,48 @@
+
+  output += words
+    .map((word, i) => {
+      const edlEntry = `${padZeros(i + 1, entries)}\tAX\tAA/V\tC`;
```
Nice!

chloebrett commented on src/sharedTypes/Word.ts:
```diff
@@ -1,7 +1,7 @@
   word: string;
 
-  start_time: number;
+  starTime: number;
```
startTime

chloebrett reviewed (changes requested):
Nearly there, still need to delete package-lock.json though. Great work taking the suggestions on board :)

There are a couple of small merge conflicts now by the way, so will need to fix those (merge develop into this branch again)

zwan0202 committed:
Merge remote-tracking branch 'origin/develop' into feature/export

chloebrett reviewed (approved):
LGTM!

patrickbrett commented:
@itsMoxMox conflicts resolved, please re check
